File: 55. Jump Game.py
# https://leetcode.com/problems/jump-game/

# Given an array of non-negative integers nums, you are initially positioned at
# the first index of the array.

# Each element in the array represents your maximum jump length at that
# position.

# Determine if you are able to reach the last index.

###############################################################################

# dp[i] = if can reach nums[i]

# greedy: start from end, determine if can move backwards

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        # return self.jump_dp(nums)
        return self.greedy(nums)
            
    def jump_dp(self, nums):
        n = len(nums)
        if n == 1: return True

        # init
        dp = [False] * n
        dp[0] = True
        
        for i in range(n):
            jump = nums[i]
            if dp[i]: # jump forwards
                if i + jump >= n - 1: # reach the end
                    return True
                else:
                    dp[i + 1 : i + jump + 1] = [True] * jump
            else:
                return False    

    def greedy(self, nums):
        n = len(nums)
        if n == 1: return True
        
        frontier = n - 1       
        for i in range(n - 1, -1, -1): # move backwards
            if i + nums[i] >= frontier: # determine if can jump to frontier
                frontier = i
        
        return frontier == 0 # determine if can reach the beginning

File: 55. Jump Game.cpp
// https://leetcode.com/problems/jump-game/

// Given an array of non-negative integers nums, you are initially positioned at
// the first index of the array.

// Each element in the array represents your maximum jump length at that position.

// Determine if you are able to reach the last index.

////////////////////////////////////////////////////////////////////////////////

// dp[i] = if can reach index i
// greedy: update frontier from end, return frontier == 0
class Solution {
public:
    bool canJump(vector<int>& nums) {
        return canJumpGreedy(nums);
        if (nums.size() == 1) return true;
        int n = nums.size();
        // dp[i] = if can reach index i
        vector<bool> dp(n, false); dp[0] = true;
        for (int i = 0; i < n - 1; ++i) {
            int jump = nums[i];
            if (dp[i]) { // can jump forward
                if (i + jump >= n - 1) return true; // reach end
                for (int j = i + 1; j <= i + jump; ++j) dp[j] = true;
            } else return false;
        }
        return false;
    }
    
    bool canJumpGreedy(vector<int>& nums) {
        if (nums.size() == 1) return true;
        int n = nums.size();
        // start from back
        int frontier = n - 1;
        for (int i = n - 2; i >= 0; --i) {
            if (i + nums[i] >= frontier) frontier = i;
        }
        return frontier == 0;
    }
};
